#### Student Name: Mai Ngo
#### Course Name and Number: DSC 430 Python Programming
#### Assignment Name and Number: Assignment0502_Cups and Dice
#### Date: 2/6/2023
#### Honor Statement: “I have not given or received any unauthorized assistance on this assignment."

#### Assumption:

 Still use top-down design. \
 I tried to code def wish Exit() separately - user decides to continue or not. It did not workout... and I gave up cause it was almost 3 hours and I still couldn't figure it out. \
 I tested all 5 sessions in the same window because I want to keeo track of the final balance. \
 Not sure the format of session verification. So I do it at the end.

In [78]:
!pip install ipynb

In [97]:
from ipynb.fs.defs.Assignment0501 import Cup

In [400]:
import random
def cupDices():
    name = whatsUp()
    #1.Greet the user and ask their name.
    
    wallet = 100
    #2.Provide the user with a balance of 100 dollars.
    
    choice = input('Do you want to play?? Press any key to continue, or "X" to exit:')
    #3.Ask if they would like to play a game. 
    
    while choice not in 'Xx':
    #While loop in case user decide to keep playing the game.
    
        goal_num = random.randrange (1, 101)
        #4.Generate a random number between 1 and 100.
        
        print (f'''Your goal number is: {goal_num}.''')
        #Print goal number so user can make their bet decision.
        
        bet_money = float(eval(input("How much you would like to bet?")))
        #5.Ask user how much they would like to bet. 
        
        if wallet > bet_money:
        #Only moving forward if user has enough money to play (bet money less than what they have).
        
            six_side, ten_side, twenty_side = askInput()
            #Assign 3 variables for quantity of each type of dice.
            #6. clarification
            
            result = cupRoll(six_side, ten_side, twenty_side)
            #Assign variable to store result from rolling all dices.
            #7.8. clarifications.
            
            wallet = myBalance(result, bet_money, goal_num,wallet)
            #Assign variable to store updated balance after a game. 
            #9.10.11. clarifications
            
            print(f'''Result: {result} | {name}'s updated balance: {wallet}.''')
            #12. Report the result to user. Message include their name and updated balance.
            
            choice = input('Do you want to continue?? Press any key to continue, or "X" to exit:') 
            #13. Ask if they would like to play again.
        else: 
        #Case when balance is not enough to bet. Stop the game.
            print("You don't have enough money. THE END. \nThank you for joining!")
            #Explain to user not enough money.
            break
        
    else: print("Sad to see you leave early. \nThank you for joining!")
    #Leave early as they wish by typing 'x' or 'X'.

In [ ]:
cupDices()

Welcome to Cups and Dices!


Please tell us your name: Mai


Your balance is: 100 dollars


Do you want to play?? Press any key to continue, or "X" to exit: m


Your goal number is: 85.


How much you would like to bet? 10
Please enter number(s) of 6 sided-roll:  4
Please enter number(s) of 10 sided-roll:  4
Please enter number(s) of 20 sided-roll:  4


Result: 102 | Mai's updated balance: 90.0.


Do you want to continue?? Press any key to continue, or "X" to exit: a


Your goal number is: 4.


How much you would like to bet? 10
Please enter number(s) of 6 sided-roll:  2
Please enter number(s) of 10 sided-roll:  0
Please enter number(s) of 20 sided-roll:  0


Result: 4 | Mai's updated balance: 190.0.


Do you want to continue?? Press any key to continue, or "X" to exit: i


Your goal number is: 23.


How much you would like to bet? 20
Please enter number(s) of 6 sided-roll:  0
Please enter number(s) of 10 sided-roll:  0
Please enter number(s) of 20 sided-roll:  3


Result: 51 | Mai's updated balance: 170.0.


Do you want to continue?? Press any key to continue, or "X" to exit: l


Your goal number is: 68.


How much you would like to bet? 20
Please enter number(s) of 6 sided-roll:  0
Please enter number(s) of 10 sided-roll:  10
Please enter number(s) of 20 sided-roll:  0


Result: 34 | Mai's updated balance: 150.0.


Do you want to continue?? Press any key to continue, or "X" to exit: u


Your goal number is: 62.


How much you would like to bet? 20
Please enter number(s) of 6 sided-roll:  3
Please enter number(s) of 10 sided-roll:  3
Please enter number(s) of 20 sided-roll:  2


Result: 42 | Mai's updated balance: 130.0.


Do you want to continue?? Press any key to continue, or "X" to exit: k


Your goal number is: 24.


How much you would like to bet? 15.5
Please enter number(s) of 6 sided-roll:  2
Please enter number(s) of 10 sided-roll:  3
Please enter number(s) of 20 sided-roll:  4


Result: 55 | Mai's updated balance: 114.5.


Do you want to continue?? Press any key to continue, or "X" to exit: e


Your goal number is: 52.


How much you would like to bet? 115


You don't have enough money. THE END. 
Thank you for joining!


In [401]:
def askInput(): 
    '''Getting quantity of each type of dices.'''
    
    six_side = int(eval(input("Please enter number(s) of 6 sided-roll: ")))
    ten_side = int(eval(input("Please enter number(s) of 10 sided-roll: ")))
    twenty_side = int(eval(input("Please enter number(s) of 20 sided-roll: ")))
    #6.Ask the user how many of each dice they would like to roll.
    
    return six_side, ten_side, twenty_side

In [402]:
def whatsUp():
    '''Greeting and asking users for their name.'''
    
    print("Welcome to Cups and Dices!")
    name = input("Please tell us your name:")
    #Asking user to input their name. 
    
    return name

In [394]:
def cupRoll(six_side, ten_side, twenty_side):
    '''Get sum result from rolling all the dices.'''
    
    cup = Cup(six_side, ten_side, twenty_side)
    #7. Create a cup filled with dice according to user's input.
    cup.roll()
    #rolling the cup.
    result = cup.getSum()
    #Retrieve sum value of the cup through def getSum()
    
    return result

In [403]:
def myBalance(result, bet_money, goal_num, wallet):
    '''Report final balance after each game.'''
    
    bet_money = float(bet_money)
    #Using float incase user bet "123.5"
    #Include in session example
    
    if result == goal_num:
    #9.If the roll exactly match the goal.
        wallet += bet_money * 10
        #User receives 10 times bet added to their balance.
        
    elif (goal_num - 3) <= result and result <= (goal_num + 3):
    #10.If the roll is within 3 of the goal but not over.
        wallet += bet_money * 5
        #User receives 5 times bet added to their balance.
        
    elif (goal_num - 10) <= result and result <= (goal_num + 10):
    #10.If the roll is within 3 of the goal but not over.
        wallet += bet_money * 2  
        #User receives 2 times bet added to their balance.
        
    else: 
    #If nothing satisfies, user lose their bet money.
        wallet -= bet_money
        #Subtracting bet money from starting balance.
        
    return wallet
    #Return final balance.

### Verification / Comment of each session

#### Session 1
Welcome to Cups and Dices! \
Please tell us your name: Mai \
Your balance is: 100 dollars \
Do you want to play?? Press any key to continue, or "X" to exit: m 

##### Give random goal number, and ask user for bet amount.
Your goal number is: 85.
How much you would like to bet? 10

##### cup = Cup(4, 4, 4)
Please enter number(s) of 6 sided-roll:  4 \
Please enter number(s) of 10 sided-roll:  4 \
Please enter number(s) of 20 sided-roll:  4 

##### Result value doesn't satisified any conditions. User lose bet.
##### Balance - bet money = 100 - 10 = 90
Result: 102 | Mai's updated balance: 90.0.

#### Session 2
##### User decide continue to play. Starting balance 90.
Do you want to continue?? Press any key to continue, or "X" to exit: a

##### Give random goal number, and ask user for bet amount.
Your goal number is: 4. How much you would like to bet? 10

##### cup = Cup(2, 0, 0) - Only 6-sided dice in cup.
Please enter number(s) of 6 sided-roll:  2 \
Please enter number(s) of 10 sided-roll:  0 \
Please enter number(s) of 20 sided-roll:  0

##### Result value equals to goal number. User receives 10 times bet added to their balance.
##### Balance + (10 * 10) = 90 + 100 = 190
Result: 4 | Mai's updated balance: 190.0.

#### Session 3
##### User decide continue to play. Starting balance 190.
Do you want to continue?? Press any key to continue, or "X" to exit: i

##### Give random goal number, and ask user for bet amount.
Your goal number is: 23. How much you would like to bet? 20

##### cup = Cup(0, 0, 3) - Only 20-sided dice in cup.
Please enter number(s) of 6 sided-roll:  0 \
Please enter number(s) of 10 sided-roll:  0 \
Please enter number(s) of 20 sided-roll:  3

##### Result value doesn't satisfied any conditions. User lose bet.
##### Balance - bet money = 190 - 20 = 170
Result: 51 | Mai's updated balance: 170.0.

#### Session 4
##### User decide continue to play. Starting balance 170.
Do you want to continue?? Press any key to continue, or "X" to exit: l

##### Give random goal number, and ask user for bet amount.
Your goal number is: 68. How much you would like to bet? 20

##### cup = Cup(0, 10, 3) - Only 10-sided dice in cup.
Please enter number(s) of 6 sided-roll:  0
Please enter number(s) of 10 sided-roll:  10
Please enter number(s) of 20 sided-roll:  0

##### Result value doesn't satisfied any conditions. User lose bet.
##### Balance - bet money = 170 - 20 = 150
Result: 34 | Mai's updated balance: 150.0.

#### Session 5
##### User decide continue to play. Starting balance 150.
Do you want to continue?? Press any key to continue, or "X" to exit: u

##### Give random goal number, and ask user for bet amount.
Your goal number is: 62. How much you would like to bet? 20

##### cup = Cup(3, 3, 2) 
Please enter number(s) of 6 sided-roll:  3 \
Please enter number(s) of 10 sided-roll:  3 \
Please enter number(s) of 20 sided-roll:  2

##### Result value doesn't satisfied any conditions. User lose bet.
##### Balance - bet money = 150 - 20 = 130
Result: 42 | Mai's updated balance: 130.0.

#### Session 6
##### User decide continue to play. Starting balance 130.
Do you want to continue?? Press any key to continue, or "X" to exit: k

##### Give random goal number, and ask user for bet amount. Bet amount is NOT an integer.
Your goal number is: 24. How much you would like to bet? 15.5

##### cup = Cup(2, 3, 4)
Please enter number(s) of 6 sided-roll:  2
Please enter number(s) of 10 sided-roll:  3
Please enter number(s) of 20 sided-roll:  4

##### Result value doesn't satisfied any conditions. User lose bet.
##### Balance - bet money = 130 - 15.5 = 114.5
Result: 55 | Mai's updated balance: 114.5.

##### User decide continue to play. Starting balance 114.5
Do you want to continue?? Press any key to continue, or "X" to exit: e

##### Give bet amount greater than balance. 
Your goal number is: 52. How much you would like to bet? 115

##### Not enough money, game automatically ends.
You don't have enough money. THE END. 
Thank you for joining!